File: Lecture 1/HelloWorldApp/app/src/main/java/com/codingblocks/demoApp/MainActivity.java
package com.codingblocks.demoApp;

import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Toast;

public class MainActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        Log.i("HelloWorldApp","OnCreate");


    }

    @Override
    protected void onResume() {
        super.onResume();
        Log.i("HelloWorldApp ","OnResume");
    }

    @Override
    protected void onStart() {
        super.onStart();
        Log.i("HelloWorldApp","OnStart");

    }

    @Override
    protected void onStop() {
        super.onStop();
        Log.i("HelloWorldApp","onStop");

    }

    @Override
    protected void onDestroy(){
        super.onDestroy();
        Log.i("HelloWorldApp","OnDestroy");


    }

    @Override
    protected void onRestart(){
        super.onRestart();
        Log.i("HelloWorldApp","OnRestart");


    }

    @Override
    protected void onPause(){
        super.onPause();
        Log.i("HelloWorldApp","onPause");


    }

    public void clickButton(View view) {
        Toast.makeText(this, "Hello", Toast.LENGTH_LONG).show();
    }
}

File: Lecture19/CameraX/app/src/main/java/com/codingblocks/camerax/MainActivity.kt
package com.codingblocks.camerax

import android.content.pm.PackageManager
import android.graphics.Matrix
import android.os.Bundle
import android.util.Rational
import android.util.Size
import android.view.Surface
import android.view.ViewGroup
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.camera.core.*
import androidx.core.app.ActivityCompat
import kotlinx.android.synthetic.main.activity_main.*
import java.io.File

class MainActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        if (ActivityCompat.checkSelfPermission(
                this,
                android.Manifest.permission.CAMERA
            ) == PackageManager.PERMISSION_DENIED
        ) {
            ActivityCompat.requestPermissions(
                this,
                arrayOf(android.Manifest.permission.CAMERA),
                123
            )
        } else {
            textureView.post {
                startCamera()
            }
        }
    }

    private fun startCamera() {

        //to capture the image of a button click
        val imageCaptureConfig = ImageCaptureConfig.Builder().apply {
            setTargetAspectRatio(Rational(1,1))
            setCaptureMode(ImageCapture.CaptureMode.MIN_LATENCY)
        }.build()

        val imageCapture = ImageCapture(imageCaptureConfig)
        imageButton.setOnClickListener {
            val file = File(externalMediaDirs.first(),"${System.currentTimeMillis()}.jpg")

            imageCapture.takePicture(file,object : ImageCapture.OnImageSavedListener{
                override fun onImageSaved(file: File) {
                    Toast.makeText(this@MainActivity,"Photo Captured ${file.absolutePath}",Toast.LENGTH_LONG).show()
                }

                override fun onError(useCaseError: ImageCapture.UseCaseError, message: String, cause: Throwable?) {
                    Toast.makeText(this@MainActivity,"Error Capturing $message",Toast.LENGTH_LONG).show()
                }

            })
        }

        //To get the preview of camera
        val previewConfig = PreviewConfig.Builder().apply {
            setTargetResolution(Size(1080, 1080))
            setTargetAspectRatio(Rational(1, 1))
            setLensFacing(CameraX.LensFacing.BACK) //to open Rear camera
        }.build()

        val preview = Preview(previewConfig)

        preview.setOnPreviewOutputUpdateListener {
            val parent = textureView.parent as ViewGroup
            parent.removeView(textureView)
            parent.addView(textureView,0)
            updatePreview()
            textureView.surfaceTexture = it.surfaceTexture
        }
        CameraX.bindToLifecycle(this, preview,imageCapture)

    }
    private fun updatePreview(){
        val matrix = Matrix()

        val centerX = textureView.width/2f
        val centerY = textureView.height/2f

        val rotation = when(textureView.display.rotation){
            Surface.ROTATION_0 -> 0
            Surface.ROTATION_90 -> 90
            Surface.ROTATION_180 -> 180
            Surface.ROTATION_270 -> 270
            else -> return
        }

        matrix.postRotate(-rotation.toFloat(),centerX,centerY)

        textureView.setTransform(matrix)
    }
}
